src/app/extensions/store-locator/pages/store-locator/store-locator-page.component.ts

```diff
@@ -0,0 +1,118 @@
+import { FormGroup } from '@angular/forms';
+import { FormlyFieldConfig } from '@ngx-formly/core';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+import { AppFacade } from 'ish-core/facades/app.facade';
+import { Country } from 'ish-core/models/country/country.model';
+
+import { StoreLocatorFacade } from '../../facades/store-locator.facade';
+import { StoreHelper } from '../../models/store/store.helper';
+import { Store as StoreModel } from '../../models/store/store.model';
+import { StoresMapComponent } from '../../shared/stores-map/stores-map.component';
+
+@Component({
+  selector: 'ish-store-locator-page',
+  templateUrl: './store-locator-page.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class StoreLocatorPageComponent implements OnInit {
+  @ViewChild('map') map: StoresMapComponent;
+
+  loaded$: Observable<boolean>;
+  stores$: Observable<StoreModel[]>;
+  highlightedStore$: Observable<StoreModel>;
+
+  form = new FormGroup({});
+  model = { countryCode: '', postalCode: '', city: '' };
+  fields: FormlyFieldConfig[];
+
+  constructor(private storeLocatorFacade: StoreLocatorFacade, private appFacade: AppFacade) {}
+
+  ngOnInit() {
+    this.fields = this.getFields();
+    this.loaded$ = this.storeLocatorFacade.getLoaded$();
+    this.stores$ = this.storeLocatorFacade.getStores$();
+    this.highlightedStore$ = this.storeLocatorFacade.getHighlighted$();
+  }
+  private getFields() {
+    return [
+      {
+        key: 'countryCode',
+        type: 'ish-select-field',
+        wrappers: ['form-field-horizontal'],
+        templateOptions: {
+          required: false,
+          label: 'store_locator.form.country.label',
+          placeholder: 'store_locator.form.country.default',
+          options: this.appFacade
+            .countries$()
+            ?.pipe(
+              map((countries: Country[]) =>
+                countries.map(country => ({ value: country.countryCode, label: country.name }))
+              )
+            ),
+        },
+      },
+      {
+        key: 'postalCode',
+        type: 'ish-text-input-field',
+        templateOptions: {
+          type: 'text',
+          label: 'store_locator.form.postalCode.label',
+        },
+      },
+      {
+        key: 'city',
+        type: 'ish-text-input-field',
+        templateOptions: {
+          type: 'text',
+          label: 'store_locator.form.city.label',
+        },
+      },
+    ];
+  }
+
+  submitForm() {
+    const countryCode = this.model.countryCode;
+    const postalCode = this.model.postalCode;
+    const city = this.model.city;
+
+    this.storeLocatorFacade.loadStores(countryCode, postalCode, city);
+  }
+
+  empty(stores: StoreModel[]) {
+    return !stores?.length;
+  }
+
+  highlighted(store: StoreModel, highlight: StoreModel) {
+    return StoreHelper.equal(store, highlight);
+  }
+
+  scrollTo(container: HTMLElement, element: HTMLElement) {
```

Comment:
Sorry if this is a dumb question, I'm not 100% sure about the setup here, but have you made sure that this function can never be called by the SSR container? Using `window` doesn't work in SSR so sometimes you have to guard these function calls with `isPlatformBrowser`.

Reply:
`scrollTo` gets called the first time after user searches for stores and stores were loaded (after `loadStores` action gets dispatched). So it shouldn't get called during SSR and only after user actually interacts with the page

Reply:
Sounds reasonable. Do you think there might be a better location to put the scrolling logic? It isn't really component-specific after all and very reusable...I was thinking of creating a scrollService or at least some static helper methods in `src/app/core/utils`.
Another question: Why don't you just use `element.scrollIntoView()` to control the scrolling instead of the complicated, manual way? Is it that important to be able to control the speed and easing granularly. `scrollIntoView` also supports easing :)

Reply:
Extracting it to a service sounds reasonable.

As to your second question, other than solving the smoothing in safari, this custom method also doesn't scroll the whole page down but only the specified scrollable container. I found if somewhat disturbing if the page itself scrolls which is why I wanted to keep it only to the container. Other than that, `scrollIntoView` with smoothing creates a weird bug when clicking on search more than once, where the list scrolls all the way to the end for no reason. The custom method was created to deal with that problem in the first place.

Reply:
Makes sense. The scrolling was more out of curiosity, the service extraction is what I think is important :D So please do that and good luck with the scrolling solution 👍